### PR TITLE
Translog: close channel on failures while converting a writer to a reader

### PR DESCRIPTION
**Pull request by bleskes**
TranslogWriter.closeIntoReader transfers the file ownership from a writer to a reader and closes the writer. If the transfer fails, we need to make sure we closed the underlying channel as the writer is already closes.

See: http://build-us-00.elastic.co/job/es_core_master_regression/4355

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 1]
```diff
@@ -275,6 +286,7 @@ protected final void ensureOpen() {
     @Override
     public final void close() throws IOException {
         if (closed.compareAndSet(false, true)) {
+            logger.trace("closing channel on writer close. path [{}]", path);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: is this really needed we already have sooo much logging everywhere

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 2]
```diff
@@ -192,7 +192,18 @@ public synchronized TranslogReader closeIntoReader() throws IOException {
...
         }
         if (closed.compareAndSet(false, true)) {
-            return new TranslogReader(generation, channel, path, firstOperationOffset, getWrittenOffset(), operationCounter);
+            logger.trace("closing into a reader. path [{}]", path);
+            try {
+                return new TranslogReader(generation, channel, path, firstOperationOffset, getWrittenOffset(), operationCounter, shardId);
+            } catch (Throwable t) {
+                // close the channel, as we are closed and failed to create a new reader
+                try {
+                    channel.close();
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: just use closeWhileHandlingException?

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java [hunk 3]
```diff
@@ -64,11 +63,11 @@
 
 
     public TranslogWriter(ShardId shardId, long generation, FileChannel channel, Path path, ByteSizeValue bufferSize) throws IOException {
-        super(generation, channel, path, channel.position());
-        this.shardId = shardId;
+        super(generation, channel, path, channel.position(), shardId);
         this.outputStream = new BufferedChannelOutputStream(java.nio.channels.Channels.newOutputStream(channel), bufferSize.bytesAsInt());
         this.lastSyncedOffset = channel.position();
         totalOffset = lastSyncedOffset;
+        logger.trace("writer created. path [{}]", path);
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: I think the logging in this class is unneeded. if you debug it you can add it back

File: core/src/main/java/org/elasticsearch/index/translog/TranslogReader.java [hunk 4]
```diff
@@ -158,6 +160,7 @@ public Checkpoint getInfo() {
     @Override
     public final void close() throws IOException {
         if (closed.compareAndSet(false, true)) {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogReader.java]: same here please remvoe the unnecessary logging

**Comment by s1monw** [comment]
left some comments especially regarding logging

File: core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java (hunk 2 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java]: there is a valid recent criticism about it swallowing exceptions and I wanted see where things come from, but yeah, can change it here and wait for the bigger change on closeWhileHandlingExceptions semantics

**Comment by bleskes** [comment]
@s1monw I pushed the change to closeWhileHandlingException . Regarding the logging - I personally find them invaluable in tracing this kind of issues. They are all trace logs and therefore shouldn't be any disturbance until you need them and turn them on. That upside greatly outweighs, at least for the way I work, the extra lines in the code.

**Comment by s1monw** [comment]
> @s1monw I pushed the change to closeWhileHandlingException . Regarding the logging - I personally find them invaluable in tracing this kind of issues. They are all trace logs and therefore shouldn't be any disturbance until you need them and turn them on. That upside greatly outweighs, at least for the way I work, the extra lines in the code.

I managed to keep this part of the code clean from these unnecessary code lines. I really don't wanna bloat it up with logger dependencies and shard ids what have you. Really please don't add them everywhere we suffer from the massive logging problem all over the place.

**Comment by s1monw** [comment]
can we please get this bug fixed with the minimal changes needed here I'd really appreciate it.

**Comment by GlenRSmith** [comment]
Without diving too deep into the code here - assuming the logging threshold is set to TRACE for sufficient scope, would it be possible to identify the shard at these code points without actually logging the shard_id in these functions/methods whose signatures must be changed exclusively for logging?

**Comment by s1monw** [comment]
@bleskes this test case fails locally for me can you remove all the unnecessary bloat here and push this. I will wait an hour then push a fix myself.

**Pull request closed without merging**